dev-tools/omdb/src/bin/omdb/db/saga.rs

```diff
@@ -101,49 +117,174 @@ async fn cmd_sagas_running(
+                            // not be contacted.
+                            //
+                            // Note: this could be seen if Nexus is up but
+                            // unreachable from where omdb is run!
+                        }
+                    },
+                }
+            }
+        }
+    }
+
+    // Find all the nodes where there is a started record but not a done record
+
+    let started_nodes: Vec<SagaNodeEvent> = {
+        use db::schema::saga_node_event::dsl;
+
+        dsl::saga_node_event
+            .filter(dsl::saga_id.eq(args.saga_id))
+            .filter(dsl::event_type.eq(SagaNodeEventType::Started.label()))
+            .load_async(&*conn)
+            .await?
+    };
+
+    let complete_nodes: Vec<SagaNodeEvent> = {
         use db::schema::saga_node_event::dsl;
 
+        // Note the actual enum contents don't matter in both these cases, it
+        // won't affect the label string
+        let succeeded_label =
+            SagaNodeEventType::Succeeded(Arc::new(serde_json::Value::Null))
+                .label();
+
+        let failed_label =
+            SagaNodeEventType::Failed(ActionError::InjectedError).label();
+
         dsl::saga_node_event
             .filter(dsl::saga_id.eq(args.saga_id))
-            .order(dsl::event_time.desc())
-            .limit(1)
-            .first_async(&*conn)
+            .filter(
+                dsl::event_type
+                    .eq(succeeded_label)
+                    .or(dsl::event_type.eq(failed_label)),
+            )
+            .load_async(&*conn)
             .await?
     };
 
-    // Inject a fault for that node, which will cause the saga to unwind
-    let action_error = steno::ActionError::action_failed(String::from(
-        "error injected with omdb",
-    ));
-
-    let fault = SagaNodeEvent {
-        saga_id: most_recent_node.saga_id,
-        node_id: most_recent_node.node_id,
-        event_type: steno::SagaNodeEventType::Failed(action_error.clone())
-            .label()
-            .to_string(),
-        data: Some(serde_json::to_value(action_error)?),
-        event_time: chrono::Utc::now(),
-        creator: most_recent_node.creator,
+    let incomplete_nodes: HashSet<u32> = {
+        let started_node_ids: HashSet<u32> =
+            started_nodes.iter().map(|node| node.node_id.0.into()).collect();
+        let complete_node_ids: HashSet<u32> =
+            complete_nodes.iter().map(|node| node.node_id.0.into()).collect();
+
+        started_node_ids.difference(&complete_node_ids).cloned().collect()
     };
 
-    {
-        use db::schema::saga_node_event::dsl;
+    let incomplete_nodes: Vec<&SagaNodeEvent> = {
+        let mut result = vec![];
+
+        for node_id in incomplete_nodes {
+            let Some(node) = started_nodes
+                .iter()
+                .find(|node| node.node_id.0 == node_id.into())
+            else {
+                bail!("could not find node?");
```

Comment:
Should this be a panic?  When would this ever happen?

Reply:
I don't think so. 2df69c2e8 changes this to unwrap (plus a comment explaining why I don't think so haha)